segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogReader.java

```diff
@@ -201,15 +207,40 @@ public String toString() {
                 // Empty ledger;
                 this.reader = null;
             }
+        }
+
+        boolean isEmpty() {
+            return this.reader == null;
+        }
 
+        private void close() {
+            // Release memory held by BookKeeper internals.
+            // we have to prevent a double free
+            if (closed.compareAndSet(false, true)) {
+
+                if (currentLedgerEntries != null) {
+                    currentLedgerEntries.close();
+                }
+                // closing a ReadHandle is mostly a no-op, it is not expected
+                // to really fail
+                try {
+                    Ledgers.close(handle);
+                } catch (DurableDataLogException bkEx) {
+                    log.error("Unable to close ReadHandle for Ledger {}.", handle.getId(), bkEx);
+                }
+            }
         }
 
         @SneakyThrows(BKException.class)
         private Iterator<LedgerEntry> readRange(long fromEntryId, long toEntryId) {
-            return Iterators.forEnumeration(Exceptions.handleInterruptedCall(() -> this.handle.readEntries(fromEntryId, toEntryId)));
+            if (currentLedgerEntries != null) {
```

Comment:
Unfortunately we must keep a reference to the current **LedgerEntries** in order to release memory